### PR TITLE
[runtime][hip][cuda] Fix semaphore multi-wait, action GPU events and cleanup

### PR DESCRIPTION
**Pull request by sogartar**
* Fix a bug where we may miss a signal when waiting on multiple semaphores. Improve code reusability by increasing code sharing in multi-wait and single-wait.

* Fix destruction in error path to not leak actions.

* Fix acquisition of action GPU events from semaphore timepoints. The switch to using iree_hal_*_event_t instead of HIP/CUDA events directly, introduced by 459fab66b0c59389b88c7bb8a8bfe81492581d9b, caused mishandling of timepoint events.
This may cause erroneous removal of action wait events and leaking them.

* Add a mutex-protected counter in the work area that keeps track of all pending host stream callbacks.
We need to make sure that all work is done before exiting the worker and then tear down the execution context.

### CODE REVIEW

**Comment by sogartar** [comment]
These are a few fixes here that I can try to disentangle if it is too hard to follow.
The CI tests may fail then.

**Comment by sogartar** [comment]
I am not sure if we want this by default when TSan is enabled since there is a problem in some HAL backend.

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c [hunk 1]
```diff
@@ -343,35 +373,26 @@ iree_status_t iree_hal_hip_semaphore_multi_wait(
...
-            semaphore, semaphore_list.payload_values[i], timeout, &timepoint);
-        if (iree_status_is_ok(status)) {
-          timepoints[timepoint_count++] = timepoint;
-          status =
-              iree_wait_set_insert(wait_set, timepoint->timepoint.host_wait);
+      iree_hal_hip_timepoint_t* timepoint;
+      status = iree_hal_hip_semaphore_try_wait_or_acquire_wait_timepoint(
+          semaphore_list.semaphores[i], semaphore_list.payload_values[i],
+          timeout, &timepoint);
+      if (!iree_status_is_ok(status)) {
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/event_semaphore.c]: Can just drop the `{ .. }` to make it oneliner.

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c [hunk 2]
```diff
@@ -102,20 +102,29 @@ static void iree_hal_hip_semaphore_destroy(
   IREE_TRACE_ZONE_END(z0);
 }
 
-static iree_status_t iree_hal_hip_semaphore_query(
+static iree_status_t iree_hal_hip_semaphore_query_locked(
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/event_semaphore.c]: This is only called in `iree_hal_hip_semaphore_query`? Why split out?

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c [hunk 3]
```diff
@@ -255,9 +264,13 @@ bool iree_hal_hip_semaphore_acquire_event_host_wait(
   return *out_event != NULL;
 }
 
-static iree_status_t iree_hal_hip_semaphore_wait(
+// Check if the semaphore has to wait to reach `value`.
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/event_semaphore.c]: Checks..

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c [hunk 4]
```diff
@@ -255,9 +264,13 @@ bool iree_hal_hip_semaphore_acquire_event_host_wait(
...
+// If it has to wait, then acquire a wait timepoint and return it.
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/event_semaphore.c]: .. returns ..

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 5]
```diff
@@ -189,6 +198,31 @@ IREE_TYPED_ATOMIC_SLIST_WRAPPER(iree_hal_hip_ready_action,
...
+  }
+  iree_hal_hip_ready_action_slist_deinitialize(list);
+}
+
+static iree_hal_hip_queue_action_t* iree_hal_hip_atomic_slist_entry_pop_front(
+    iree_hal_hip_atomic_slist_entry_t* list) {
+  IREE_ASSERT(list->ready_list_head);
+
+  iree_hal_hip_queue_action_t* action = list->ready_list_head;
+  if (action->next) {
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c:219]: Can drop `{ .. }` to be oneliner.

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 6]
```diff
@@ -189,6 +198,31 @@ IREE_TYPED_ATOMIC_SLIST_WRAPPER(iree_hal_hip_ready_action,
...
+    action->next->prev = NULL;
+  }
+  list->ready_list_head = action->next;
+
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c:223]: Should we clear the current `action`'s `next` before returning?

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 7]
```diff
@@ -218,14 +252,22 @@ typedef struct iree_hal_hip_working_area_t {
...
   iree_notification_t exit_notification;
   iree_hal_hip_ready_action_slist_t ready_worklist;  // atomic
   iree_atomic_int32_t worker_state;                  // atomic
+  // TODO: use status to provide more context for the error.
   iree_atomic_intptr_t error_code;                   // atomic
   // The number of actions that have been issued to the GPU but not yet fully
   // completed both execution and cleanup. We don't need this field to be atomic
   // given it is modified only from the worker thread.
   int32_t pending_action_count;
+
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: Add some docs on this? Explain what's the difference with `pending_action_count`? I think it's worth to name them according to the lifetime a bit, maybe like `action_cleanup_to_complete_count`, `action_cleanup_to_issue_count`?

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 8]
```diff
@@ -433,6 +484,32 @@ static void iree_hal_hip_free_semaphore_list(
   iree_allocator_free(host_allocator, semaphore_list->payload_values);
 }
 
+void iree_hal_hip_queue_action_destroy(iree_hal_hip_queue_action_t* action) {
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: `static`?

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 9]
```diff
@@ -556,6 +639,19 @@ static void iree_hal_hip_execution_device_signal_host_callback(
...
 
+  iree_status_t status;
+
+  // Need to signal the list before zombifying the action, because in the mean
+  // time someone else may issue the pending queue actions.
+  // If we push first to the pending actions list, the cleanup of this action
+  // may run while we are still using the semaphore list, causing a crash.
+  status = iree_hal_semaphore_list_signal(action->signal_semaphore_list);
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+    IREE_ASSERT(false && "can't signal semaphores");
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: "cannot signal semaphores in host callback"

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 10]
```diff
@@ -565,24 +661,27 @@ static void iree_hal_hip_execution_device_signal_host_callback(
...
-    // of the recently zombified action so it can be cleaned up.
-    status = iree_hal_hip_pending_queue_actions_issue(actions);
-  }
-
+  // We need to trigger execution of this action again, so it gets cleaned up.
+  status = iree_hal_hip_pending_queue_actions_issue(actions);
   if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
-    IREE_ASSERT(false &&
-                "can't signal semaphores and/or issue pending actions");
+    IREE_ASSERT(false && "can't issue pending actions");
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: "cannot issue action for cleanup in host callback"

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 11]
```diff
@@ -845,10 +919,6 @@ iree_status_t iree_hal_hip_pending_queue_actions_issue(
   }
 
   if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
-    // Release all actions in the ready list to avoid leaking.
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c:848]: Why is this deleted? Is this related to your later TODO of leaking resources? Can we put some comment to explain here too?

**Review by antiagainst — CHANGES_REQUESTED** [review]
Thanks Boian! A few comments. Though I think it might be better to grab a time to chat about this.

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c (hunk 1 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/event_semaphore.c]: Done

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c (hunk 2 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/event_semaphore.c]: A remnant from a previous attempt. I removed it.

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c (hunk 3 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/event_semaphore.c]: done

File: runtime/src/iree/hal/drivers/hip/event_semaphore.c (hunk 4 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/event_semaphore.c]: done

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 5 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c:219]: done

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 6 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c:223]: Good catch. I fixed it.

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 7 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: Do you mean that `pending_action_count` should instead mean only up to cleanup? It currently means including cleanup.
`pending_action_count` -> `action_cleanup_to_complete_count`
`pending_action_count` becomes the mutex protected counter that is decremented in the host stream callback.

I added some comments. I chose this name since it exactly says what we are counting. The other code will also not need to change if we ever add other host stream callbacks.

There is one other approach to join the 2 counters into one mutex protected counter that is more generic like `pending_work_items_count`. But then we would need to a lock at the end of action cleanup to decrease the counter. Now we don't need synchronization as the `pending_action_count` is only modified/read from the worker thread. The worker thread exit logic would be a bit simpler if we have to care about only one counter. On the other hand it will be harder to disentangle counting bugs if we ever add other kinds of work items.

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 8 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: done

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 9 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: done

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 10 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: done

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 11 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c:848]: My bad. Yes, this would introduce a leak. I added the destruction of the list.

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 7 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: I merged the two counters.

**Comment by sogartar** [comment]
@antiagainst, could you review again?

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c [hunk 12]
```diff
@@ -218,14 +253,23 @@ typedef struct iree_hal_hip_working_area_t {
...
-  iree_atomic_intptr_t error_code;                   // atomic
-  // The number of actions that have been issued to the GPU but not yet fully
-  // completed both execution and cleanup. We don't need this field to be atomic
-  // given it is modified only from the worker thread.
-  int32_t pending_action_count;
+  // TODO: use status to provide more context for the error.
+  iree_atomic_intptr_t error_code;  // atomic
+
+  // The number asynchroneous work items that are scheduled and not complete.
+  // For example
```
Comment (antiagainst) [inline runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: I think we can be exact here regarding the lifetime of each action--can mention that for each action we have exactly two async work items so it's very clear. "For example" typically means non exclusive.

**Review by antiagainst — APPROVED** [review]
Looks good! Thanks a lot for working through these issues! I just have one more comment regarding documentation. Also don't forget to mirror to cuda side.

File: runtime/src/iree/hal/drivers/hip/pending_queue_actions.c (hunk 12 above)
Comment (sogartar) [reply runtime/src/iree/hal/drivers/hip/pending_queue_actions.c]: done

**Comment by sogartar** [comment]
Thank you for the review. I rebased and mirrored the change to the CUDA driver.

**Comment by sogartar** [comment]
Rebased to get the fix from ea7d01e4279bab94e1fdb5b0ade9a82f20b6c32d.
